source/common/uma_pools/disjoint_pool.cpp

```diff
@@ -0,0 +1,948 @@
+    // empty slabs within the Available list to speed up the process of checking
+    // if a slab in this bucket is already pooled.
+    size_t chunkedSlabsInPool;
+
+    // Statistics
+    size_t allocPoolCount;
+    size_t freeCount;
+    size_t currSlabsInUse;
+    size_t currSlabsInPool;
+    size_t maxSlabsInPool;
+
+  public:
+    // Statistics
+    size_t allocCount;
+    size_t maxSlabsInUse;
+
+    Bucket(size_t Sz, DisjointPool::AllocImpl &AllocCtx)
+        : Size{Sz}, OwnAllocCtx{AllocCtx}, chunkedSlabsInPool(0),
+          allocPoolCount(0), freeCount(0), currSlabsInUse(0),
+          currSlabsInPool(0), maxSlabsInPool(0), allocCount(0),
+          maxSlabsInUse(0) {}
+
+    // Get pointer to allocation that is one piece of an available slab in this
+    // bucket.
+    void *getChunk(bool &FromPool);
+
+    // Get pointer to allocation that is a full slab in this bucket.
+    void *getSlab(bool &FromPool);
+
+    // Return the allocation size of this bucket.
+    size_t getSize() const { return Size; }
+
+    // Free an allocation that is one piece of a slab in this bucket.
+    void freeChunk(void *Ptr, Slab &Slab, bool &ToPool);
+
+    // Free an allocation that is a full slab in this bucket.
+    void freeSlab(Slab &Slab, bool &ToPool);
+
+    uma_memory_provider_handle_t getMemHandle();
+
+    DisjointPool::AllocImpl &getAllocCtx() { return OwnAllocCtx; }
+
+    // Check whether an allocation to be freed can be placed in the pool.
+    bool CanPool(bool &ToPool);
+
+    // The minimum allocation size for any slab.
+    size_t SlabMinSize();
+
+    // The allocation size for a slab in this bucket.
+    size_t SlabAllocSize();
+
+    // The minimum size of a chunk from this bucket's slabs.
+    size_t ChunkCutOff();
+
+    // The number of slabs in this bucket that can be in the pool.
+    size_t Capacity();
+
+    // The maximum allocation size subject to pooling.
+    size_t MaxPoolableSize();
+
+    // Update allocation count
+    void countAlloc(bool FromPool);
+
+    // Update free count
+    void countFree();
+
+    // Update statistics of Available/Unavailable
+    void updateStats(int InUse, int InPool);
+
+    // Print bucket statistics
+    void printStats(bool &TitlePrinted, const std::string &Label);
+
+  private:
+    void onFreeChunk(Slab &, bool &ToPool);
+
+    // Update statistics of pool usage, and indicate that an allocation was made
+    // from the pool.
+    void decrementPool(bool &FromPool);
+
+    // Get a slab to be used for chunked allocations.
+    decltype(AvailableSlabs.begin()) getAvailSlab(bool &FromPool);
+
+    // Get a slab that will be used as a whole for a single allocation.
+    decltype(AvailableSlabs.begin()) getAvailFullSlab(bool &FromPool);
+};
+
+class DisjointPool::AllocImpl {
+    // It's important for the map to be destroyed last after buckets and their
+    // slabs This is because slab's destructor removes the object from the map.
```

Comment:
In that case, it might be a good idea to create a destructor for `AllocImpl` that destroys every slab in `KnownSlabs` before its lifecycle ends.